### PR TITLE
[Need review] Added scrollable arrangements for issue #346

### PR DESCRIPTION
**Pull request by shumelchyk**
Fix for issue #346 about scrollable arrangements.

Here is my instance to test changes:
http://web-downloader.appspot.com/

And link to the companion apk:
https://www.dropbox.com/s/nlbe5oqopv74kid/companion.apk?dl=0

Both horizontal and vertical arrangements are fixed, here is small demo:
https://www.dropbox.com/s/cegcoofsvr0f4zq/scroll-views.mov?dl=0

Scrolls will work in final apk file.

### CODE REVIEW

**Comment by puravidaapps** [comment]
great! looking forward to this feature...
unfortunately I was not able to test this, the companion app looks like this after starting.
screenshot_20160529-092102

 Also trying to connect (with code or also using QR code) unfortunately did not work, it just does not happen anything... Trying to build an example app also does not work, because no build server was available...

Taifun

**Comment by halatmit** [comment]
@puravidaapps 
To test this from the pull request, you need to build an instance on your local machine and also build a companion.  Have you done that sort of thing before?    I'll be looking at this PR later today.

**Comment by puravidaapps** [comment]
thank you @halatmit
I tried to test using the mentioned instance at http://web-downloader.appspot.com/
together with the provided companion apk at https://www.dropbox.com/s/nlbe5oqopv74kid/companion.apk?dl=0

**Comment by halatmit** [comment]
@shumelchyk 
I've not done a complete review, but you need to disable the Center and Bottom choices in the AlignVertical dropdown when the vertical arrangement is set  Scrollable.   Do this in MockHVArrangement.  Look at the analogous code in MockForm, which has special code relating to scrollable in onPropertyChange and enableAndDisableDropdowns.   You might want to also do something similar for horizontal arrangements, but I'm not sure it's worth it.

Another issue is the lack of a scrollbar on the scrollable vertical arrangement in the designer.   You'll see that there's a scrollbar on phone, and there's  scrollbar on the designer in the case of Screens.  You can see the code for this scrollbar in MackForm.   But it's very involved, and my advice is that it's not worth the effort for now.   I'd rather you finish quickly and move on to your next issue.

So fix the dropdown, and don't do the scrollbar.   Update the pull request and I'll continue the review.

**Comment by shumelchyk** [comment]
Thanks @halatmit for comments. I am going to fix the choices.

**Comment by shumelchyk** [comment]
Disabled bottom and center choices for AlignVertical dropdown when the arrangement is scrollable.
Updated this PR.
Here is the test instance with latest changes:
http://web-downloader.appspot.com/

And link to the companion apk is here:
https://www.dropbox.com/s/pqux45v5xu62bo1/MIT%20AI2%20Companion.apk?dl=0

**Comment by halatmit** [comment]
@shumelchyk 
You disabled the vertical alignment for scrollable horizontal arrangements.    Good!  But now you need to do the corresponding disabling  of the horizontal alignment options for scrollable vertical arrangements.   You can see the need for this if you try it on the phone: create a vertical arrangement whose width is the width of the screen.  place a button in that vertical arrangement.  make the vertical arrangement scrollable. now  in the designer, make the vertical arrangement's horizontal alignment centered.   You'should see that the button does not move on the phone, but stays to the left.  So you need to make the designer act the same, by disabling the horizontal alignment options for center and right.   You do this in a similar way as you did for the horizontal arrangement.  But one complication is that when you are enabling and disabling the property editor in MockHVArrangement, you'll need to know whether you should deal the horizontalArrangment choices or the vertical arrangement choices.  See if you can figure out how to do that.

**Comment by shumelchyk** [comment]
Done the fix for horizontal arrangement. Commit is already added to this PR.

Here is the test instance with latest changes:
http://web-downloader.appspot.com/

And link to the companion apk is here:
https://www.dropbox.com/s/pqux45v5xu62bo1/MIT%20AI2%20Companion.apk?dl=0

**Comment by halatmit** [comment]
@shumelchyk I made a few more comments and sent them to you in email.

**Comment by shumelchyk** [comment]
Committed final changes after code review.

**Comment by halatmit** [comment]
@jisqyv @shumelchyk 
Jeff, could you take a look at this?  I can't test it while I'm on vacation.  Last I looked at it, it looked OK, and all that it needed was to update YaVersion and friends.

File: appinventor/components/src/com/google/appinventor/components/common/YaVersion.java [hunk 1]
```diff
@@ -680,11 +680,12 @@ private YaVersion() {
 
   public static final int GYROSCOPESENSOR_COMPONENT_VERSION = 1;
 
-  // For HORIZONTALARRANGEMENT_COMPONENT_VERSION 2:
+  // For HORIZONTALARRANGEMENT_COMPONENT_VERSION 3:
+  // - The scrollable property was added
```
Comment (halatmit) [inline appinventor/components/src/com/google/appinventor/components/common/YaVersion.java]: The comments are not quite right.   AlignHorizontal and AlignVertical and color (I think) were part of the upgrade to version 3.   Then scrollabe is part of the upgrade to version 4.
One reason you got confused is that the previous version of YaVersion had a bug.   See my email to you and to Jeff Schiller.

File: appinventor/components/src/com/google/appinventor/components/common/YaVersion.java [hunk 2]
```diff
@@ -938,11 +939,12 @@ private YaVersion() {
   //   TwitPic. The TwitPic_API_Key property is now deprecated and hidden.
   public static final int TWITTER_COMPONENT_VERSION = 4;
 
-  // For VERTICALARRANGEMENT_COMPONENT_VERSION 2:
+  // For VERTICALARRANGEMENT_COMPONENT_VERSION 3:
```
Comment (halatmit) [inline appinventor/components/src/com/google/appinventor/components/common/YaVersion.java]: See comment above to HorizontalArrangement.

Comment (halatmit) [inline appinventor/components/src/com/google/appinventor/components/common/YaVersion.java]: Also, you need to add lines earlier in the file that say

  // For YOUNG_ANDROID_VERSION ???:
  // - HORIZONTALARRANGEMENT_COMPONENT_VERSION was incremented to 4.
  // - VERTICALARRANGEMENT_COMPONENT_VERSION was incremented to 4

File: appinventor/components/src/com/google/appinventor/components/common/YaVersion.java [hunk 3]
```diff
@@ -680,11 +680,12 @@ private YaVersion() {
 
   public static final int GYROSCOPESENSOR_COMPONENT_VERSION = 1;
 
-  // For HORIZONTALARRANGEMENT_COMPONENT_VERSION 2:
+  // For HORIZONTALARRANGEMENT_COMPONENT_VERSION 3:
+  // - The scrollable property was added
   // - The AlignHorizontal property was added
   // - The AlignVertical property was added
```
Comment (jisqyv) [inline appinventor/components/src/com/google/appinventor/components/common/YaVersion.java]: Before the line "// - Added background color & image" we need to add:
    // For HORIZONTALARRANGEMENT_COMPONENT_VERSION 3:
This is to fix a bug from a previous change (when we added the background color and image, this line was left out.
The similar change needs to be added for VERITICALARRANGEMENT_COMPONENT_VERSION as well.

This is *not* your issue, but given you are updating this, it is a good time to correct the problem. Thanks.

-Jeff

**Comment by shumelchyk** [comment]
I fixed YaVersion file. @halatmit @jisqyv please check it

**Comment by jisqyv** [comment]
@shumelchyk The merge conflict is in YaVersion. Since you started working on this change other changes have been merged into the system. You can fix the conflict by merging the "ucr" branch into your change and fix the conflict. Note: "ucr" stands for "Upcoming Component Release" which is the branch we put changes in that effect the components. Prior to a component release, we merge it into master. At the moment master and ucr are on the same commit.

I'm in the process of reviewing and testing your change (you should test too :-) ). When it is ready I will "squash" it down to one commit that will go on the ucr branch.

-Jeff

**Comment by jisqyv** [comment]
@shumelchyk Not quite ready. I get compilation errors when in YoungAndroidFormUpgrader:

AiClientLib:
    [mkdir] Created dir: /u1/jis/app-inventor/appinventor/appengine/build/war/WEB-INF/classes
    [javac] Compiling 370 source files to /u1/jis/app-inventor/appinventor/appengine/build/war/WEB-INF/classes
    [javac] /u1/jis/app-inventor/appinventor/appengine/src/com/google/appinventor/client/youngandroid/YoungAndroidFormUpgrader.java:947: error: illegal start of expression
    [javac]     }, 
    [javac]      ^
    [javac] 1 error

I was able to build the Companion, but I'm seeing the same broken UI that Taifun was seeing:
screenshot

**Comment by shumelchyk** [comment]
@jisqyv Thanks, I already fixed this bug - it was redundant symbol ',' after the brackets.

**Comment by shumelchyk** [comment]
@jisqyv @halatmit I have merged latest changes from 'ucr' branch and resolved conflicts in YaVersion, looking forward to your review

**Comment by jisqyv** [comment]
@shumelchyk I see you did the merge successfully and also fixed the problem in YoungAndroidFormUprader that caused it to not compile. However you still need to address the UI strangeness in the Companion App (see screen shots above).

**Comment by halatmit** [comment]
@shumelchyk @jisqyv @afmckinney 
I just checked, and the companion UI bug is still in this PR.   It is not in the master branch.

Jeff, this is likely to be some small glitch, but it might be hard for Kseniia to find, given that it's no obvious how the companion is created.    Could take a look and give her a hint -- or just do a git compare and find the bug?

**Comment by jisqyv** [comment]
@shumelchyk @halatmit I believe the problem here is that you are creating a new ScrollView instead of using the already provided layout manager. This is probably OK for "Responsive" mode, but in "Fixed" mode (which is the default mode of the Companion) you likely have to embed the ScrollView in a ScaledLayout and scale it accordingly. There is code in Form.java that does this, it might be worth a look.

**Comment by jisqyv** [comment]
@shumelchyk Another thing to look at is whether you can splice your ScrollView into the existing view hierarchy. Then it might scale when the App is scaled.

-Jeff

**Comment by jisqyv** [comment]
@shumelchyk You might need to use a device (or emulator with a density > 1.0 in order to see what is happening).

**Comment by halatmit** [comment]
@shumelchyk @jisqyv See if you can make progress based on Jeff's comments.   This is turning out to be trickier that I thought, but try to push it through.  Remember Jeff's warning to test this on screens of different densities, and also with both fixed and responsive scaling.   Let me know if you want to talk before Friday.

**Comment by jisqyv** [comment]
@shumelchyk @halatmit So I've been hacking around with your change. It turns out the HorizontalScrollView is causing the borkage. Here are the changes I have made over where you are. This code is by no means finished (we have to handle enabling/disabling scrolling).

https://github.com/jisqyv/appinventor-sources/commit/bb39564bb5e0ecb523e836d7a3416f8664cde42f

-Jeff

**Comment by halatmit** [comment]
@jisqyv @shumelchyk This had been turned over to Jeff to finish, with Hal.

**Comment by jisqyv** [comment]
@shumelchyk @halatmit A revised version of this change has been committed to the UCR (Upcoming Component Release) branch. You can see the resulting work here:

https://github.com/mit-cml/appinventor-sources/commit/d621adfd82e12fef649a4968d6394d2c352e33ba

Reviewed on Gerrit and approved by Andrew

**Pull request closed without merging**